runner/javaut.go

```diff
@@ -1,36 +1,129 @@
-	// response, err := jr.Start(ctx, w, r, codeTask)
-	// if err != nil {
-	// 	return http.StatusInternalServerError, err
-	// }
-	//jr.Handle(ctx, w, response, resultKey)
-	return nil
-}
+func JUnit(ctx context.Context, params map[string]string, sub model.KeyedSubmission) (stdout, stderr *bytes.Buffer, testResults []model.UnitTestResults, err error) {
+	image := newImage("javaut")
+
+	if err = prepareImage(ctx, image); err != nil {
+		return nil, nil, []model.UnitTestResults{}, err
+	}
+
+	var v *docker.Volume
+	if v, err = createDockerVolume(sub.Code.Bucket + "/" + path.Dir(sub.Code.Name)); err != nil {
+		return
+	}
+
+	var testV *docker.Volume
+	if testV, err = createDockerVolume(params["tests"]); err != nil {
+		return
+	}
+
+	var c *docker.Container
+	c, err = dc.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image: image,
+		},
+		HostConfig: &docker.HostConfig{
+			Privileged: false,
+			Memory:     0, // TODO(flowlo): Limit memory
+			Binds:      []string{v.Name + ":/run/src/main/java", testV.Name + ":/run/src/test/java"},
+		},
+	})
+	if err != nil {
+		return
+	}
+
+	start := time.Now()
+	if err = dc.StartContainer(c.ID, c.HostConfig); err != nil {
+		return
+	}
+
+	waitc := make(chan waitResult)
+	go func() {
+		exitCode, err := dc.WaitContainer(c.ID)
+		waitc <- waitResult{exitCode, err}
+	}()
 
-//
-// // Start function for a java unit test run.
-// func (jr *JunitRunner) Start(ctx context.Context, w http.ResponseWriter, r *http.Request, codeTask model.CodeTask) (response *http.Response, err error) {
-// 	if err = json.NewDecoder(r.Body).Decode(&jr.Submission); err != nil {
-// 		return
-// 	}
-// 	// TODO(victorbalan): pass the correct language when we have different JUnit tests
-// 	return run(codeTask, "javaut", jr.Submission.Code)
-// }
-//
-// // Handle function for a JUnit test run.
-// func (jr *JunitRunner) Handle(ctx context.Context, w http.ResponseWriter, response *http.Response, resultKey *datastore.Key) {
-// 	json.NewDecoder(response.Body).Decode(&jr.Submission)
-//
-// 	key, err := jr.Submission.SaveWithParent(ctx, resultKey)
-// 	if err != nil {
-// 		return
-// 	}
-// 	json.NewEncoder(w).Encode(jr.Submission.Key(key))
-// }
+	var res waitResult
+	select {
+	case res = <-waitc:
+	case <-time.After(time.Minute):
+		return nil, nil, []model.UnitTestResults{}, errors.New("execution timed out")
+	}
+
+	if res.Err != nil {
+		return nil, nil, []model.UnitTestResults{}, res.Err
+	}
+
+	stdout, stderr = new(bytes.Buffer), new(bytes.Buffer)
+	err = dc.Logs(docker.LogsOptions{
+		Container:    c.ID,
+		OutputStream: stdout,
+		ErrorStream:  stderr,
+		Stdout:       true,
+		Stderr:       true,
+	})
+	if err != nil {
+		return
+	}
+	var buf bytes.Buffer
+	err = dc.CopyFromContainer(docker.CopyFromContainerOptions{
+		Container:    c.ID,
+		Resource:     params["resultPath"],
+		OutputStream: &buf,
+	})
+	if err != nil {
+		return
+	}
+
+	var utr model.UnitTestResults
```

Comment:
Move this declaration as close to it's usage possible. (One line above `xml.Unmarshal`)